File: tools/extended_devices/door_valve.py
from __future__ import annotations

from ..base import BaseTool, ToolResult
from ..schemas.extended_devices import GarageDoorInput, ValveInput


def _extract_error_message(data: dict) -> str:
    error_detail = data.get("error", {})
    if isinstance(error_detail, dict):
        return error_detail.get("message", str(error_detail))
    return str(error_detail)


class GarageDoorTool(BaseTool):
    name = "garage_door_control"
    description = "차고문을 열거나 닫습니다."
    args_schema = GarageDoorInput

    def execute(self, device_id: str, action: str) -> ToolResult:
        data = self.client.command(device_id, "doorControl", action)
        if "error" in data:
            return ToolResult(success=False, error=_extract_error_message(data))
        return ToolResult(
            success=True,
            data={"device_id": device_id, "action": action, "result": data},
        )


class ValveTool(BaseTool):
    name = "valve_control"
    description = "스마트 밸브(수도 차단 밸브 등)를 열거나 닫습니다."
    args_schema = ValveInput

    def execute(self, device_id: str, action: str) -> ToolResult:
        data = self.client.command(device_id, "valve", action)
        if "error" in data:
            return ToolResult(success=False, error=_extract_error_message(data))
        return ToolResult(
            success=True,
            data={"device_id": device_id, "action": action, "result": data},
        )

File: smartthings_tools_rs/src/extended_devices/door_valve.rs
use schemars::schema::RootSchema;
use schemars::schema_for;
use serde_json::Value;

use crate::client::SmartThingsClient;
use crate::common_tools::{extract_error_msg, has_error};
use crate::schemas::extended_devices::*;
use crate::tool::{Tool, ToolResult};

pub struct GarageDoorTool<'a> {
    pub client: &'a SmartThingsClient,
}

impl Tool for GarageDoorTool<'_> {
    fn name(&self) -> &'static str {
        "garage_door_control"
    }
    fn description(&self) -> &'static str {
        "차고문을 열거나 닫습니다."
    }
    fn parameters_schema(&self) -> RootSchema {
        schema_for!(GarageDoorInput)
    }

    fn execute(&self, args: Value) -> ToolResult {
        let input: GarageDoorInput = match serde_json::from_value(args) {
            Ok(v) => v,
            Err(e) => return ToolResult::fail(format!("ValidationError: {e}")),
        };
        let data = self
            .client
            .command(&input.device_id, "doorControl", &input.action, None, None);
        if has_error(&data) {
            return ToolResult::fail(extract_error_msg(&data));
        }
        ToolResult::ok(
            serde_json::json!({ "device_id": input.device_id, "action": input.action, "result": data }),
        )
    }
}

pub struct ValveTool<'a> {
    pub client: &'a SmartThingsClient,
}

impl Tool for ValveTool<'_> {
    fn name(&self) -> &'static str {
        "valve_control"
    }
    fn description(&self) -> &'static str {
        "스마트 밸브(수도 차단 밸브 등)를 열거나 닫습니다."
    }
    fn parameters_schema(&self) -> RootSchema {
        schema_for!(ValveInput)
    }

    fn execute(&self, args: Value) -> ToolResult {
        let input: ValveInput = match serde_json::from_value(args) {
            Ok(v) => v,
            Err(e) => return ToolResult::fail(format!("ValidationError: {e}")),
        };
        let data = self
            .client
            .command(&input.device_id, "valve", &input.action, None, None);
        if has_error(&data) {
            return ToolResult::fail(extract_error_msg(&data));
        }
        ToolResult::ok(
            serde_json::json!({ "device_id": input.device_id, "action": input.action, "result": data }),
        )
    }
}
